docs: oidc client assertions and pkce (#25375)
hashicorp/nomad: website/content/api-docs/acl/auth-methods.mdx

        - "client_secret": Use the `OIDCClientSecret` inherited from the parent
          `ACLAuthMethodConfig` as an HMAC key

    - `KeyAlgorithm` `(string)` is the key's algorithm. Its default values are
      based on the `KeySource`:
        - "nomad": "RS256" (from Nomad's keyring, must not be changed)
        - "private_key": "RS256" (must be RS256, RS384, or RS512)
        - "client_secret": "HS256" (must be HS256, HS384, or HS512)

    - `PrivateKey` `(OIDCClientAssertionKey)` - External key material to sign
      the JWT. `KeySource` must be "private_key" to enable this.

      - `PemKey` `(string)` - An RSA private key, in pem format. It is used to
        sign the JWT. Mutually exclusive with `PemKeyFile`.

      - `PemKeyFile` `(string)` - An absolute path to a private key on Nomad
        servers' disk, in pem format. It is used to sign the JWT.
        Mutually exclusive with `PemKey`.

      - `KeyIDHeader` `(string)` - Which header the provider will use to find
        the public key to verify the signed JWT.
        Allowed values are: "kid", "x5t", or "x5t#S256".
        The default and allowed values depend on which of the following options

>         is set.

Author: I like where you're going with it, but that's a bit wordy for me, and the precision in the examples ironically ends up distracting me a bit.  How about

> \- `KeyIDHeader` `(string)` - Which header the provider uses to find the public key to verify the signed JWT. The default and allowed values depend on whether you set `KeyID`,  `PemCert`, or `PemCertFile`. You must set exactly one of those options, so refer to them for their requirements.

      - `KeyID` `(string)` - Becomes the JWT's "kid" header.
        Mutually exclusive with `PemCert` and `PemCertFile`.
        Allowed `KeyIDHeader` values: "kid" (the default)

      - `PemCert` `(string)` - An x509 certificate, signed by the private key
        or a CA, in pem format. It is used to derive an [x5t#S256][]
        (or [x5t][]) KeyID.
        Mutually exclusive with `PemCertFile` and `KeyID`.
        Allowed `KeyIDHeader` values: "x5t", "x5t#S256" (default "x5t#S256")

      - `PemCertFile` `(string)` - An absolute path to an x509 certificate on
        Nomad servers' disk, signed by the private key or a CA, in pem format.

>         It is used to derive an [x5t#S256][] (or [x5t][]) KeyID.

Reviewer: ```suggestion
        Nomad uses this certificate to derive an [x5t#S256][] or [x5t][] KeyID.
```

        Mutually exclusive with `PemCert` and KeyID.
        Allowed `KeyIDHeader` values: "x5t", "x5t#S256" (default "x5t#S256")

    - `ExtraHeaders` `(map[string]string)` - Added to the JWT headers,
      alongside "kid" and "type". Setting the "kid" header here is not allowed;
      use `PrivateKey.KeyID`.

>   - `OIDCDisablePKCE` `(bool: false)` - When set to `true`, Nomad will not

Reviewer: ```suggestion
  - `OIDCDisablePKCE` `(bool: false)` - When set to `true`, Nomad does not
```

    include [PKCE][] verification in the auth flow. Even with PKCE enabled in

>     Nomad (the default), you may still need to enable it in the OIDC provider.

Reviewer: ```suggestion
    Nomad, which is the default, you may still need to enable PKCE in the OIDC provider.
```

        - "private_key": Use key material in the `PrivateKey` field
        - "client_secret": Use the `OIDCClientSecret` inherited from the parent
          `ACLAuthMethodConfig` as an HMAC key

    - `KeyAlgorithm` `(string)` is the key's algorithm. Its default values are
      based on the `KeySource`:
        - "nomad": "RS256" (from Nomad's keyring, must not be changed)
        - "private_key": "RS256" (must be RS256, RS384, or RS512)
        - "client_secret": "HS256" (must be HS256, HS384, or HS512)

    - `PrivateKey` `(OIDCClientAssertionKey)` - External key material to sign
      the JWT. `KeySource` must be "private_key" to enable this.

      - `PemKey` `(string)` - An RSA private key, in pem format. It is used to
        sign the JWT. Mutually exclusive with `PemKeyFile`.

      - `PemKeyFile` `(string)` - An absolute path to a private key on Nomad
        servers' disk, in pem format. It is used to sign the JWT.
        Mutually exclusive with `PemKey`.

      - `KeyIDHeader` `(string)` - Which header the provider will use to find
        the public key to verify the signed JWT.
        Allowed values are: "kid", "x5t", or "x5t#S256".

>         The default depends on which of the following options is set.

Reviewer 2: ```suggestion
        The default and allowed values depend on which of the following options is set.
```

  - `OIDCClientSecret` `(string: "")` - The OAuth client secret
    configured with your OIDC provider.

>   - `OIDCClientAssertion` `(OIDCClientAssertion)` - Optionally send a signed JWT

Reviewer: Same comments as above. At first I thought the parameters under Create and Update were identical and was going to suggest putting them in a partial so we don't copy/paste content... but then I noticed a few Create parameters that aren't in Update so there goes the partial idea at this time. I might address this later, when I'm updating front matter descriptions for search.

I'd prefer the API docs be generated from the code, and I think Education is moving in that direction.

Reviewer: Could still put the common parameters in a partial, since the partial would just be a markdown list. 

These parameters are in Create but not Update:
- JWTValidationPubKeys
- JWKSURL
- JWKSCACert
- ExpirationLeeway
- NotBeforeLeeway
- ClockSkewLeeway
- VerboseLogging

So common parms in partial and those 7 after the included partial in Create. 
Also has me wondering if those 7 parms are supposed to be in update but docs never updated?? Hmmmm

Author: It would not surprise me one bit if they should actually be identical, but someone missed updating them because it's easy to get wrong! I'll double-check on this.

Author: Alrighty, I checked, and those params work with Update, too, so I think they were just an oversight.

db3e0eb179e92dad5440d77a851c219ec030ea62 makes the whole block into a partial.
